### ISSUE TITLE
Out of order messages across cycles

### ISSUE THREAD

**Issue by RobAustin** [OpenHFT/Chronicle-Queue#498]
@JerryShea  this was captured when running the net.openhft.chronicle.queue.impl.single.RollCycleMultiThreadStressTest

This was from DumpQueueMain.dump(`<file>`), it contains 2 queue files the first is the oldest but it contains the highest value [ it should contain the lowest ]

@JerryShea @peter-lawrey -- regarding this issue, @dpisklov and I have been investigating it. 

We feel its possible at locking the writing document context during a write may not be sufficient to guarantee the order of writes, especially if the appenders [ from many threads ] where created just as the queue rolls. The implicit contract is that within a  try (DocumentContext dc = writingDocument()) { } no other appender may write to any other .cq4 file, regardless of which .cq4 file its writing to. This appears from tests at least, not to always be the case [ however we don't exactly have definitive proof yet but this queue dump would indicate it ]

One option is to use the table store to lock the queue. when ever any appender writes to it, regardless of the CQ4 being written to, it would also have to support recovery [ for appenders that get killed during a write ]

```
Writing 66666 messages with 30000ns interval
Should take ~0sec
[pos: 524822, rlim: 524839, wlim: 140737488355328, cap: 140737488355328 ] ٠٠٠٠٠٠٠٠٠٠٠٠٠٠٠٠٠٠٠٠٠٠٠٠٠٠٠٠٠٠٠٠٠٠٠٠٠٠٠٠٠٠٠٠٠٠٠٠٠٠\u001B٠٠٠§°xòBJ\u0017٠٠⒈ǁ⒈⒈⒈⒈⒈⒈⒈⒈⒈⒈⒈⒈⒈⒈⒈⒈⒈‡٠٠٠À٠٠٠٠
Expected: 0, actual: 1, pos: 0, index: 200000000, cycle: 2
[pos: 524822, rlim: 524839, wlim: 140737488355328, cap: 140737488355328 ] ٠٠٠٠٠٠٠٠٠٠٠٠٠٠٠٠٠٠٠٠٠٠٠٠٠٠٠٠٠٠٠٠٠٠٠٠٠٠٠٠٠٠٠٠٠٠٠٠٠٠\u001B٠٠٠§°xòBJ\u0017٠٠⒈ǁ⒈⒈⒈⒈⒈⒈⒈⒈⒈⒈⒈⒈⒈⒈⒈⒈⒈‡٠٠٠À٠٠٠٠
Expected: 0, actual: 1, pos: 0, index: 200000000, cycle: 2
--- !!meta-data #binary
header: !SCQStore {
  wireType: !WireType BINARY_LIGHT,
  writePosition: [
    524808,
    2254033196679168
  ],
  roll: !SCQSRoll {
    length: !short 1000,
    format: yyyyMMdd-HHmmss,
    epoch: 0
  },
  indexing: !SCQSIndexing {
    indexCount: !int 32768,
    indexSpacing: 4,
    index2Index: 450,
    lastIndex: 4
  },
  lastAcknowledgedIndexReplicated: -1,
  recovery: !TimedStoreRecovery {
    timeStamp: 0
  },
  deltaCheckpointInterval: 0,
  lastIndexReplicated: -1,
  sourceId: 0
}
# position: 450, header: -1
--- !!meta-data #binary
index2index: [
  # length: 32768, used: 1
  262632,
  
]
# position: 524808, header: 0
--- !!data #binary
25607718205616
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
# position: 524839, header: 0 EOF
--- !!not-ready-meta-data! #binary
...
# 785877 bytes remaining
--- !!meta-data #binary
header: !SCQStore {
  wireType: !WireType BINARY_LIGHT,
  writePosition: [
    524808,
    2254033196679168
  ],
  roll: !SCQSRoll {
    length: !short 1000,
    format: yyyyMMdd-HHmmss,
    epoch: 0
  },
  indexing: !SCQSIndexing {
    indexCount: !int 32768,
    indexSpacing: 4,
    index2Index: 450,
    lastIndex: 4
  },
  lastAcknowledgedIndexReplicated: -1,
  recovery: !TimedStoreRecovery {
    timeStamp: 0
  },
  deltaCheckpointInterval: 0,
  lastIndexReplicated: -1,
  sourceId: 0
}
# position: 450, header: -1
--- !!meta-data #binary
index2index: [
  # length: 32768, used: 1
  262632,
  ...
]
# position: 262632, header: -1
--- !!meta-data #binary
index: [
  # length: 32768, used: 1
  524808,
 ...
]
# position: 524808, header: 0
--- !!data #binary
25607717875641
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
# position: 524839, header: 0 EOF
--- !!not-ready-meta-data! #binary
...
# 785877 bytes remaining
```

**Comment by JerryShea** [comment]
See also net.openhft.chronicle.queue.impl.single.DocumentOrderingTest, one of which is @Ignored(flaky)

**Comment by JerryShea** [comment]
This should be resolved by https://github.com/OpenHFT/Chronicle-Queue/issues/504

**Comment by dpisklov** [comment]
Resolved by #504 (should be at least, reopen if it's not)

**Comment by hft-team-city** [comment]
Released in Chronicle-Queue-5.16.0 (https://github.com/OpenHFT/Chronicle-Queue/releases/chronicle-queue-5.16.0), BOM-1.16.74 (https://github.com/OpenHFT/OpenHFT/releases/chronicle-bom-1.16.74)
